activesupport/lib/active_support/core_ext/module/attribute_accessors.rb

```diff
@@ -203,4 +203,57 @@ def mattr_accessor(*syms, instance_reader: true, instance_writer: true, instance
     mattr_writer(*syms, instance_writer: instance_writer, instance_accessor: instance_accessor, default: default, location: location)
   end
   alias :cattr_accessor :mattr_accessor
+
+  def singleton_attr_reader(*syms, instance_reader: true, instance_accessor: true, default: nil, location: nil)
+    raise TypeError, "singleton attributes should be defined directly on class, not singleton" if singleton_class?
+    location ||= caller_locations(1, 1).first
```

Comment:
Are we using `location`?

Reply:
Indeed not, since we can't use `class_eval` anymore as we need closures.